core/corehttp/gateway_handler.go

```diff
@@ -626,3 +638,42 @@ func webErrorWithCode(w http.ResponseWriter, message string, err error, code int
 func internalWebError(w http.ResponseWriter, err error) {
 	webErrorWithCode(w, "internalWebError", err, http.StatusInternalServerError)
 }
+
+func searchUpTreeFor404(ctx context.Context, acceptHeaders []string, parsedPath coreiface.Path, i *gatewayHandler) (coreiface.Path, error) {
+	filename404, err := preferred404Filename(acceptHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	pathComponents := strings.SplitAfter(gopath.Dir(parsedPath.String()), "/")
+
+	for idx := len(pathComponents); idx >= 3; idx-- {
+		pretty404 := gopath.Join(append(pathComponents[0:idx], filename404)...)
+		parsedPath, err := coreapi.ParsePath(pretty404)
+		if err == nil {
+			resolvedPath, err := i.api.ResolvePath(ctx, parsedPath)
+			if err == nil {
+				return resolvedPath, nil
+			}
+		}
+	}
```

Comment:
Working with paths will generally be simpler soon, there's a few changes to `coreiface.Path` I have in mind.